nodedb.go

```diff
@@ -532,46 +551,125 @@ func (ndb *nodeDB) DeleteVersionsRange(fromVersion, toVersion int64) error {
 		}
 	}
 
+	// notify stop the loading works
+	if ndb.cacheOrphansCancel != nil {
+		ndb.cacheOrphansCancel()
+	}
+
+	miss := 0
+	hit := 0
+	hitKeysCount := 0
+	missKeysCount := 0
+
+	t1 := time.Now()
+	totalOrphans := 0
+	totalDelete := int64(0)
 	// If the predecessor is earlier than the beginning of the lifetime, we can delete the orphan.
 	// Otherwise, we shorten its lifetime, by moving its endpoint to the predecessor version.
 	for version := fromVersion; version < toVersion; version++ {
-		err := ndb.traverseOrphansVersion(version, func(key, hash []byte) error {
+		orphans, ok := ndb.cachedOrphans.Load(version)
+		if !ok {
+			if err := ndb.loadOrphans(version); err != nil {
+				return err
+			}
+			orphans, ok = ndb.cachedOrphans.Load(version)
```

Comment:
Why trying to load from cache for the second time?

Reply:
The preceding line, ndb.loadOrphans(version), stores the retrieved orphan nodes in cachedOrphans, so reload them to orphans. Here 'cachedOrphans' serves as a temporary variable.